File: CoreBankerWeb/CoreBanker/Services/DigitalBankingService.cs
namespace CoreBanker.Services;

public class DigitalBankingService : ApiClientBase
{
    public DigitalBankingService(HttpClient httpClient, CoreBanker.State.AppState appState) : base(httpClient, appState) { }

    public Task<DigitalBankingDashboardDto?> GetDashboardAsync(CancellationToken cancellationToken = default)
        => GetAsync<DigitalBankingDashboardDto>("/api/digital-banking/dashboard", cancellationToken);

    public async Task<List<DigitalBankingProductDto>> GetSavingsProductsAsync(CancellationToken cancellationToken = default)
        => await GetAsync<List<DigitalBankingProductDto>>("/api/digital-banking/savings/products", cancellationToken) ?? [];

    public async Task<List<AccountListItemDto>> GetCustomerSavingsAccountsAsync(string customerId, CancellationToken cancellationToken = default)
        => await GetAsync<List<AccountListItemDto>>($"/api/digital-banking/savings/accounts/{Uri.EscapeDataString(customerId)}", cancellationToken) ?? [];

    public Task<AccountListItemDto?> OpenSavingsAccountAsync(OpenDigitalSavingsAccountRequest request, CancellationToken cancellationToken = default)
        => PostAsync<OpenDigitalSavingsAccountRequest, AccountListItemDto>("/api/digital-banking/savings/accounts", request, cancellationToken);

    public Task<AccountListItemDto?> FundSavingsAccountAsync(string accountId, DigitalSavingsTransferRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalSavingsTransferRequest, AccountListItemDto>($"/api/digital-banking/savings/accounts/{Uri.EscapeDataString(accountId)}/fund", request, cancellationToken);

    public Task<AccountListItemDto?> WithdrawSavingsAccountAsync(string accountId, DigitalSavingsTransferRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalSavingsTransferRequest, AccountListItemDto>($"/api/digital-banking/savings/accounts/{Uri.EscapeDataString(accountId)}/withdraw", request, cancellationToken);

    public Task<DigitalInvestmentPortfolioDto?> GetInvestmentPortfolioAsync(string? customerId = null, CancellationToken cancellationToken = default)
    {
        var url = string.IsNullOrWhiteSpace(customerId)
            ? "/api/digital-banking/investments/portfolio"
            : $"/api/digital-banking/investments/portfolio?customerId={Uri.EscapeDataString(customerId)}";
        return GetAsync<DigitalInvestmentPortfolioDto>(url, cancellationToken);
    }

    public Task<DigitalInvestmentProfileDto?> CreateInvestmentAsync(CreateDigitalInvestmentRequest request, CancellationToken cancellationToken = default)
        => PostAsync<CreateDigitalInvestmentRequest, DigitalInvestmentProfileDto>("/api/digital-banking/investments", request, cancellationToken);

    public Task<DigitalInvestmentProfileDto?> TopUpInvestmentAsync(string profileId, DigitalInvestmentActionRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalInvestmentActionRequest, DigitalInvestmentProfileDto>($"/api/digital-banking/investments/{Uri.EscapeDataString(profileId)}/top-up", request, cancellationToken);

    public Task<DigitalInvestmentProfileDto?> RolloverInvestmentAsync(string profileId, DigitalInvestmentActionRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalInvestmentActionRequest, DigitalInvestmentProfileDto>($"/api/digital-banking/investments/{Uri.EscapeDataString(profileId)}/rollover", request, cancellationToken);

    public Task<DigitalInvestmentProfileDto?> LiquidateInvestmentAsync(string profileId, DigitalInvestmentActionRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalInvestmentActionRequest, DigitalInvestmentProfileDto>($"/api/digital-banking/investments/{Uri.EscapeDataString(profileId)}/liquidate", request, cancellationToken);

    public Task<DigitalLoanEligibilityDto?> CheckLoanEligibilityAsync(CheckDigitalLoanEligibilityRequest request, CancellationToken cancellationToken = default)
        => PostAsync<CheckDigitalLoanEligibilityRequest, DigitalLoanEligibilityDto>("/api/digital-banking/loans/eligibility", request, cancellationToken);

    public Task<LoanDto?> ApplyLoanAsync(CreateDigitalLoanApplicationRequest request, CancellationToken cancellationToken = default)
        => PostAsync<CreateDigitalLoanApplicationRequest, LoanDto>("/api/digital-banking/loans/apply", request, cancellationToken);

    public Task<LoanDto?> RepayLoanAsync(string loanId, DigitalLoanRepaymentRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalLoanRepaymentRequest, LoanDto>($"/api/digital-banking/loans/{Uri.EscapeDataString(loanId)}/repay", request, cancellationToken);

    public Task<LoanDto?> RestructureLoanAsync(DigitalLoanRestructureRequest request, CancellationToken cancellationToken = default)
        => PostAsync<DigitalLoanRestructureRequest, LoanDto>("/api/digital-banking/loans/restructure", request, cancellationToken);
}

public class DigitalBankingDashboardDto
{
    public int ActiveSavingsAccounts { get; set; }
    public decimal TotalSavingsBalance { get; set; }
    public int ActiveInvestmentProfiles { get; set; }
    public decimal TotalInvestmentBalance { get; set; }
    public int ActiveLoans { get; set; }
    public decimal TotalLoanExposure { get; set; }
    public int PendingApprovals { get; set; }
}

public class DigitalBankingProductDto
{
    public string ProductCode { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string Type { get; set; } = string.Empty;
    public string Currency { get; set; } = "GHS";
    public decimal? InterestRate { get; set; }
    public string? InterestMethod { get; set; }
    public decimal? MinAmount { get; set; }
    public decimal? MaxAmount { get; set; }
    public int? DefaultTerm { get; set; }
    public string Status { get; set; } = "ACTIVE";
}

public class OpenDigitalSavingsAccountRequest
{
    public string CustomerId { get; set; } = string.Empty;
    public string ProductCode { get; set; } = string.Empty;
    public string? BranchId { get; set; }
    public string? Currency { get; set; }
    public decimal InitialDepositAmount { get; set; }
    public string? FundingAccountId { get; set; }
}

public class DigitalSavingsTransferRequest
{
    public string CounterpartyAccountId { get; set; } = string.Empty;
    public decimal Amount { get; set; }
    public string Narration { get; set; } = string.Empty;
}

public class CreateDigitalInvestmentRequest
{
    public string CustomerId { get; set; } = string.Empty;
    public string FundingAccountId { get; set; } = string.Empty;
    public string ProductCode { get; set; } = string.Empty;
    public decimal Principal { get; set; }
    public decimal Rate { get; set; }
    public int TenorDays { get; set; }
    public string PayoutOption { get; set; } = "AT_MATURITY";
    public bool AutoRollover { get; set; }
    public string? Notes { get; set; }
}

public class DigitalInvestmentActionRequest
{
    public string? FundingAccountId { get; set; }
    public decimal? Amount { get; set; }
    public DateTime? NewMaturityDate { get; set; }
    public decimal? NewRate { get; set; }
    public string? DestinationAccountId { get; set; }
    public decimal? PenaltyAmount { get; set; }
    public string? Notes { get; set; }
}

public class DigitalInvestmentProfileDto
{
    public string Id { get; set; } = string.Empty;
    public string AccountId { get; set; } = string.Empty;
    public string CustomerId { get; set; } = string.Empty;
    public string FundingAccountId { get; set; } = string.Empty;
    public string ProductCode { get; set; } = string.Empty;
    public string Currency { get; set; } = "GHS";
    public decimal Principal { get; set; }
    public decimal Rate { get; set; }
    public int TenorDays { get; set; }
    public string PayoutOption { get; set; } = "AT_MATURITY";
    public bool AutoRollover { get; set; }
    public string Status { get; set; } = "ACTIVE";
    public DateTime StartDate { get; set; }
    public DateTime MaturityDate { get; set; }
    public decimal ProjectedMaturityValue { get; set; }
    public string? Notes { get; set; }
}

public class DigitalInvestmentPortfolioDto
{
    public int ActiveProfiles { get; set; }
    public decimal TotalPrincipal { get; set; }
    public decimal TotalProjectedMaturityValue { get; set; }
    public Dictionary<string, decimal> ByCurrency { get; set; } = new();
    public List<DigitalInvestmentProfileDto> Items { get; set; } = new();
}

public class CheckDigitalLoanEligibilityRequest
{
    public string CustomerId { get; set; } = string.Empty;
    public string? LoanProductId { get; set; }
    public decimal? Principal { get; set; }
    public string? ProviderName { get; set; }
}

public class DigitalLoanEligibilityDto
{
    public bool IsEligible { get; set; }
    public List<string> Reasons { get; set; } = new();
    public CreditCheckDto CreditCheck { get; set; } = new();
}

public class CreateDigitalLoanApplicationRequest
{
    public string CustomerId { get; set; } = string.Empty;
    public string LoanProductId { get; set; } = string.Empty;
    public decimal Principal { get; set; }
    public string? ServicingAccountId { get; set; }
    public string? CollateralAccountId { get; set; }
    public string? ClientReference { get; set; }
}

public class DigitalLoanRepaymentRequest
{
    public decimal Amount { get; set; }
    public string AccountId { get; set; } = string.Empty;
    public string? ClientReference { get; set; }
}

public class DigitalLoanRestructureRequest
{
    public string LoanId { get; set; } = string.Empty;
    public int NewTermInPeriods { get; set; }
    public decimal? NewAnnualRate { get; set; }
    public string? NewRepaymentFrequency { get; set; }
    public string Reason { get; set; } = "Restructure";
}

File: src/services/digitalBankingService.ts
import { httpClient } from './httpClient';
import { API_ENDPOINTS } from './apiConfig';
import { Account, Customer, Loan } from '../../types';

export interface DigitalBankingDashboard {
  activeSavingsAccounts: number;
  totalSavingsBalance: number;
  activeInvestmentProfiles: number;
  totalInvestmentBalance: number;
  activeLoans: number;
  totalLoanExposure: number;
  pendingApprovals: number;
}

export interface DigitalBankingProduct {
  productCode: string;
  name: string;
  type: string;
  currency: string;
  interestRate?: number;
  interestMethod?: string;
  minAmount?: number;
  maxAmount?: number;
  defaultTerm?: number;
  status: string;
}

export interface DigitalInvestmentProfile {
  id: string;
  accountId: string;
  customerId: string;
  fundingAccountId: string;
  productCode: string;
  currency: string;
  principal: number;
  rate: number;
  tenorDays: number;
  payoutOption: string;
  autoRollover: boolean;
  status: string;
  startDate: string;
  maturityDate: string;
  projectedMaturityValue: number;
  notes?: string;
}

export interface DigitalInvestmentPortfolio {
  activeProfiles: number;
  totalPrincipal: number;
  totalProjectedMaturityValue: number;
  byCurrency: Record<string, number>;
  items: DigitalInvestmentProfile[];
}

export interface DigitalLoanEligibility {
  isEligible: boolean;
  reasons: string[];
  creditCheck: {
    customerId: string;
    score: number;
    internalScore?: number;
    bureauScore?: number | null;
    compositeScore?: number;
    probabilityGood?: number;
    riskBand: string;
    riskGrade: string;
    decision: string;
    recommendation: string;
    checkedAt?: string;
  };
}

export interface OpenDigitalSavingsAccountRequest {
  customerId: string;
  productCode: string;
  branchId?: string;
  currency?: string;
  initialDepositAmount?: number;
  fundingAccountId?: string;
}

export interface CreateDigitalInvestmentRequest {
  customerId: string;
  fundingAccountId: string;
  productCode: string;
  principal: number;
  rate: number;
  tenorDays: number;
  payoutOption: string;
  autoRollover?: boolean;
  notes?: string;
}

export interface DigitalSavingsTransferRequest {
  counterpartyAccountId: string;
  amount: number;
  narration: string;
}

export interface DigitalInvestmentActionRequest {
  fundingAccountId?: string;
  amount?: number;
  newMaturityDate?: string;
  newRate?: number;
  destinationAccountId?: string;
  penaltyAmount?: number;
  notes?: string;
}

export interface CheckDigitalLoanEligibilityRequest {
  customerId: string;
  loanProductId?: string;
  principal?: number;
  providerName?: string;
}

export interface CreateDigitalLoanApplicationRequest {
  customerId: string;
  loanProductId: string;
  principal: number;
  servicingAccountId?: string;
  collateralAccountId?: string;
  clientReference?: string;
}

class DigitalBankingService {
  async getDashboard(): Promise<DigitalBankingDashboard> {
    return httpClient.get<DigitalBankingDashboard>(API_ENDPOINTS.digitalBanking.dashboard);
  }

  async getSavingsProducts(): Promise<DigitalBankingProduct[]> {
    return httpClient.get<DigitalBankingProduct[]>(API_ENDPOINTS.digitalBanking.savingsProducts);
  }

  async getCustomerSavingsAccounts(customerId: string): Promise<Account[]> {
    return httpClient.get<Account[]>(API_ENDPOINTS.digitalBanking.customerSavingsAccounts(customerId));
  }

  async openSavingsAccount(data: OpenDigitalSavingsAccountRequest): Promise<Account> {
    return httpClient.post<Account>(API_ENDPOINTS.digitalBanking.openSavingsAccount, data);
  }

  async fundSavingsAccount(accountId: string, data: DigitalSavingsTransferRequest): Promise<Account> {
    return httpClient.post<Account>(API_ENDPOINTS.digitalBanking.fundSavingsAccount(accountId), data);
  }

  async withdrawSavingsAccount(accountId: string, data: DigitalSavingsTransferRequest): Promise<Account> {
    return httpClient.post<Account>(API_ENDPOINTS.digitalBanking.withdrawSavingsAccount(accountId), data);
  }

  async getInvestmentPortfolio(customerId?: string): Promise<DigitalInvestmentPortfolio> {
    const suffix = customerId ? `?customerId=${encodeURIComponent(customerId)}` : '';
    return httpClient.get<DigitalInvestmentPortfolio>(`${API_ENDPOINTS.digitalBanking.investmentPortfolio}${suffix}`);
  }

  async createInvestment(data: CreateDigitalInvestmentRequest): Promise<DigitalInvestmentProfile> {
    return httpClient.post<DigitalInvestmentProfile>(API_ENDPOINTS.digitalBanking.createInvestment, data);
  }

  async topUpInvestment(profileId: string, data: DigitalInvestmentActionRequest): Promise<DigitalInvestmentProfile> {
    return httpClient.post<DigitalInvestmentProfile>(API_ENDPOINTS.digitalBanking.topUpInvestment(profileId), data);
  }

  async rolloverInvestment(profileId: string, data: DigitalInvestmentActionRequest): Promise<DigitalInvestmentProfile> {
    return httpClient.post<DigitalInvestmentProfile>(API_ENDPOINTS.digitalBanking.rolloverInvestment(profileId), data);
  }

  async liquidateInvestment(profileId: string, data: DigitalInvestmentActionRequest): Promise<DigitalInvestmentProfile> {
    return httpClient.post<DigitalInvestmentProfile>(API_ENDPOINTS.digitalBanking.liquidateInvestment(profileId), data);
  }

  async checkLoanEligibility(data: CheckDigitalLoanEligibilityRequest): Promise<DigitalLoanEligibility> {
    return httpClient.post<DigitalLoanEligibility>(API_ENDPOINTS.digitalBanking.checkLoanEligibility, data);
  }

  async applyLoan(data: CreateDigitalLoanApplicationRequest): Promise<Loan> {
    return httpClient.post<Loan>(API_ENDPOINTS.digitalBanking.applyLoan, data);
  }

  async repayLoan(loanId: string, data: { amount: number; accountId: string; clientReference?: string }): Promise<Loan> {
    return httpClient.post<Loan>(API_ENDPOINTS.digitalBanking.repayLoan(loanId), data);
  }

  async restructureLoan(data: { loanId: string; newTermInPeriods: number; newAnnualRate?: number; newRepaymentFrequency?: string; reason: string }): Promise<Loan> {
    return httpClient.post<Loan>(API_ENDPOINTS.digitalBanking.restructureLoan, data);
  }

  async getLoanStatement(loanId: string): Promise<any> {
    return httpClient.get<any>(API_ENDPOINTS.digitalBanking.loanStatement(loanId));
  }

  async getLoanSchedule(loanId: string): Promise<any[]> {
    return httpClient.get<any[]>(API_ENDPOINTS.digitalBanking.loanSchedule(loanId));
  }
}

export const digitalBankingService = new DigitalBankingService();
